server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java

```diff
@@ -277,39 +277,45 @@ private static boolean isAggregationSort(BucketOrder order) {
         }
     }
 
-    @Override
-    protected Aggregator doCreateInternal(Aggregator parent, CardinalityUpperBound cardinality, Map<String, Object> metadata)
-        throws IOException {
-        BucketCountThresholds bucketCountThresholds = new BucketCountThresholds(this.bucketCountThresholds);
+    public static BucketCountThresholds adjustBucketCountThresholds(BucketCountThresholds bucketCountThresholds, BucketOrder order) {
+        BucketCountThresholds newBucketCountThresholds = new BucketCountThresholds(bucketCountThresholds);
         if (InternalOrder.isKeyOrder(order) == false
-            && bucketCountThresholds.getShardSize() == TermsAggregationBuilder.DEFAULT_BUCKET_COUNT_THRESHOLDS.shardSize()) {
+            && newBucketCountThresholds.getShardSize() == TermsAggregationBuilder.DEFAULT_BUCKET_COUNT_THRESHOLDS.shardSize()) {
             // The user has not made a shardSize selection. Use default
             // heuristic to avoid any wrong-ranking caused by distributed
             // counting
-            bucketCountThresholds.setShardSize(BucketUtils.suggestShardSideQueueSize(bucketCountThresholds.getRequiredSize()));
+            newBucketCountThresholds.setShardSize(BucketUtils.suggestShardSideQueueSize(newBucketCountThresholds.getRequiredSize()));
         }
+        newBucketCountThresholds.ensureValidity();
+        return newBucketCountThresholds;
+    }
+
+    @Override
+    protected Aggregator doCreateInternal(Aggregator parent, CardinalityUpperBound cardinality, Map<String, Object> metadata)
+        throws IOException {
+
+        BucketCountThresholds adjusted = adjustBucketCountThresholds(this.bucketCountThresholds, order);
```

Comment:
I think we would also adjust if no parallelism is enabled?

Reply:
Yes. My intention was to just factor the existing code out to a method that I can call from another place. I did not mean to change any of the existing logic. Have I missed anything?

Reply:
I think that the terms agg builder should just update its own shard count. Then this code doesn't need to be changed. It is not needed for other terms agg usages.

Reply:
I am not sure what you mean. that the adjustment should happen earlier in the builder? I was not sure what call may trigger that, and I think it should not be retrieving the cardinality that does that, because that's kind of an unwanted side effect of deciding whether we parallelize or not.